stacks-signer/src/signerdb.rs

```diff
@@ -826,6 +827,47 @@ impl SignerDb {
             .map(|info| serde_json::from_str(info).map_err(DBError::from))
             .collect()
     }
+
+    /// Compute the tenure extend timestamp based on the tenure start and already consumed idle time of the
+    /// globally accepted blocks of the provided tenure (identified by the cosnensus hash)
+    pub fn get_tenure_extend_timestamp(
+        &self,
+        tenure_idle_timeout: Duration,
+        consensus_hash: &ConsensusHash,
+    ) -> u64 {
+        // We do not know our tenure start timestamp until we find the last processed tenure change transaction for the given consensus hash.
+        // We may not even have it in our database, in which case, we should use the oldest known block in the tenure.
+        // If we have no blocks known for this tenure, we will assume it has only JUST started and calculate
+        // our tenure extend timestamp based on the epoch time in secs.
+        let mut tenure_start_timestamp = None;
+        let mut tenure_process_time_ms = 0;
+        // Note that the globally accepted blocks are already returned in descending order of stacks height, therefore by newest block to oldest block
+        for block_info in self
+            .get_globally_accepted_blocks(consensus_hash)
+            .unwrap_or_default()
+            .iter()
+        {
+            // Always use the oldest block as our tenure start timestamp
+            tenure_start_timestamp = Some(block_info.proposed_time);
+            tenure_process_time_ms += block_info.validation_time_ms.unwrap_or(0);
+
+            if block_info
+                .block
+                .txs
+                .first()
+                .map(|tx| matches!(tx.payload, TransactionPayload::TenureChange(_)))
+                .unwrap_or(false)
+            {
+                // Tenure change found. No more blocks should count towards this tenure's processing time.
+                break;
+            }
```

Comment:
Nit: This works but only because `get_globally_accepted_blocks` returns by block height descending. This seems like an easy gotcha. Maybe just a comment on the `get_globally_accepted_blocks` function?

Reply:
Added. Yes was very purposeful in returning in descending order so I wouldn't have to reverse iterate.